fairscale/optim/oss.py

```diff
@@ -292,10 +294,20 @@ def clip_grad_norm(
             total_norm = total_norm ** (1.0 / norm_type)
 
         clip_coef = torch.tensor(max_norm, dtype=total_norm.dtype, device=total_norm.device) / (total_norm + 1e-6)
+
+        if skip_nan_grads and torch.isnan(clip_coef):
+            return total_norm
+
+        assert not torch.isnan(clip_coef), (
+            "Clipping grads with a NaN, cannot be right.\n"
+            + f"Local norm: {local_norm.item()}. Total norm: {total_norm.item()}"
+        )
+
         if clip_coef < 1:
-            for device, device_params in self.per_device_params.items():
-                for p in filter(lambda x: x.grad is not None, device_params[self.rank]):
-                    p.grad.detach().mul_(clip_coef.to(device))  # type: ignore
+            with torch.no_grad():
```

Comment:
This seems like unrelated cleanup. With large changes it is much easier to review if there is a single thesis.

Reply:
you're right, I'll move that to another PR